File: Fundamental courses of the major/Math For Computer Science/TAI LIEU KHOA TRUOC/C++ & python/C3 Geometry/Closest pair problem 2D.py
import math

def dis(p1, p2) :
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
    #return math.sqrt((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2)

def brute_force(points) :
    min_dis = float('inf')
    n = len(points)
    for i in range (n) :
        for j in range (i + 1, n) :
            min_dis = min(min_dis, dis(points[i], points[j]))
    return min_dis

def strip_closest(strip, d) :
    min_d = d
    strip.sort(key = lambda point: point[1])
    for i in range (len(strip)) :
        for j in range (i + 1, len(strip)) :
            if abs(strip[i][1] - strip[j][1]) >= min_d :
                break
            min_d = min(min_d, dis(strip[i], strip[j]))
    return min_d

def closest(points, left, right) :
    if right - left <= 3:
        return brute_force(points[left: right + 1])
    mid = (left + right) // 2
    mid_point = points[mid]

    d_left = closest(points, left, mid)
    d_right = closest(points, mid + 1, right)

    d = min(d_left, d_right)
    strip = []
    for i in range (left, right + 1) :
        if abs(points[i][0] - mid_point[0]) < d:
            strip.append(points[i]);
    return min(d, strip_closest(strip, d))

def closest_pair(points) :
    points_sorted = sorted(points, key = lambda points: points[0])
    return closest(points_sorted, 0, len(points_sorted) - 1)

points = [(12, 41), (31, 21), (52, 17), (24, 19), (8, 19), (24, 29), (14, 27), (41, 52), (16,27), (28, 40)]
min_dis = closest_pair(points)
print(f"Khoảng cách nhỏ nhất giữa hai điểm trong tập hợp điểm trên là: {min_dis}")

File: Fundamental courses of the major/Math For Computer Science/TAI LIEU KHOA TRUOC/C++ & python/C3 Geometry/Closest pair problem 2D.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

double dis(pair<int, int> p1, pair<int, int> p2) {
    return abs(p1.first - p2.first) + abs(p1.second - p2.second);
    // return sqrt(pow(p1.first - p2.first, 2) + pow(p1.second - p2.second, 2));
}

double brute_force(const vector<pair<int, int>>& points) {
    double min_dis = numeric_limits<double>::infinity();
    int n = points.size();
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            min_dis = min(min_dis, dis(points[i], points[j]));
        }
    }
    return min_dis;
}

double strip_closest(vector<pair<int, int>>& strip, double d) {
    double min_d = d;
    sort(strip.begin(), strip.end(), [](const pair<int, int>& p1, const pair<int, int>& p2) {
        return p1.second < p2.second;
    });
    int size = strip.size();
    for (int i = 0; i < size; i++) {
        for (int j = i + 1; j < size; j++) {
            if (abs(strip[i].second - strip[j].second) >= min_d) {
                break;
            }
            min_d = min(min_d, dis(strip[i], strip[j]));
        }
    }
    return min_d;
}

double closest(const vector<pair<int, int>>& points, int left, int right) {
    if (right - left <= 3) {
        vector<pair<int, int>> subset(points.begin() + left, points.begin() + right + 1);
        return brute_force(subset);
    }
    int mid = (left + right) / 2;
    pair<int, int> mid_point = points[mid];

    double d_left = closest(points, left, mid);
    double d_right = closest(points, mid + 1, right);

    double d = min(d_left, d_right);
    vector<pair<int, int>> strip;
    for (int i = left; i <= right; i++) {
        if (abs(points[i].first - mid_point.first) < d) {
            strip.push_back(points[i]);
        }
    }
    return min(d, strip_closest(strip, d));
}

double closest_pair(vector<pair<int, int>>& points) {
    sort(points.begin(), points.end());
    return closest(points, 0, points.size() - 1);
}

int main() {
    vector<pair<int, int>> points = {{12, 41}, {31, 21}, {52, 17}, {24, 19}, {8, 19},
                                      {24, 29}, {14, 27}, {41, 52}, {16, 27}, {28, 40}};
    double min_dis = closest_pair(points);
    cout << "Khoảng cách nhỏ nhất giữa hai điểm trong tập hợp điểm trên là: " << min_dis << endl;
    return 0;
}
